"""
AIVA gRPC å®¢æˆ¶ç«¯ - V2 çµ±ä¸€æ¶æ§‹æ•´åˆ
=====================================

æ•´åˆ gRPC é€šä¿¡åˆ° AivaClient æ˜“ç”¨å±¤
æ”¯æ´èˆ‡ç¾æœ‰ HTTP/MQ å”è­°ä¸¦å­˜

åŠŸèƒ½:
- çµ±ä¸€å®¢æˆ¶ç«¯ä»‹é¢
- è‡ªå‹•å”è­°é¸æ“‡ (gRPC/HTTP/MQ)
- é€£æ¥æ± ç®¡ç†
- è‡ªå‹•é‡è©¦èˆ‡å®¹éŒ¯
- åˆ†æ•£å¼è¿½è¹¤
"""

import asyncio
import logging
import uuid
from typing import Dict, Any, Optional, AsyncIterator
from datetime import datetime
import json

import grpc
from grpc import aio  # type: ignore

# gRPC ç”Ÿæˆçš„å­˜æ ¹
try:
    from services.aiva_common.grpc.generated.python import aiva_pb2, aiva_pb2_grpc
    GRPC_AVAILABLE = True
except ImportError:
    logging.warning("gRPC å­˜æ ¹æœªæ‰¾åˆ°ï¼ŒgRPC åŠŸèƒ½å°‡è¢«ç¦ç”¨")
    aiva_pb2 = None
    aiva_pb2_grpc = None
    GRPC_AVAILABLE = False

from services.aiva_common.messaging.compatibility_layer import message_broker
from services.aiva_common.enums import Topic, ModuleName

logger = logging.getLogger(__name__)


class GRPCConnectionManager:
    """gRPC é€£æ¥ç®¡ç†å™¨"""
    
    def __init__(self):
        self._channels = {}  # é€£æ¥æ± 
        self._stubs = {}     # å­˜æ ¹ç·©å­˜
    
    async def get_channel(self, address: str) -> Optional[aio.Channel]:
        """ç²å– gRPC é€šé“"""
        if not GRPC_AVAILABLE:
            return None
            
        if address not in self._channels:
            try:
                channel = aio.insecure_channel(address)
                # æ¸¬è©¦é€£æ¥
                await channel.channel_ready()
                self._channels[address] = channel
                logger.info(f"ğŸ“¡ å»ºç«‹ gRPC é€£æ¥: {address}")
            except Exception as e:
                logger.warning(f"âš ï¸  gRPC é€£æ¥å¤±æ•—: {address} - {e}")
                return None
        
        return self._channels[address]
    
    async def get_task_stub(self, address: str):
        """ç²å–ä»»å‹™æœå‹™å­˜æ ¹"""
        if not GRPC_AVAILABLE:
            return None
            
        stub_key = f"{address}_task"
        if stub_key not in self._stubs:
            channel = await self.get_channel(address)
            if channel:
                self._stubs[stub_key] = aiva_pb2_grpc.TaskServiceStub(channel)
        
        return self._stubs.get(stub_key)
    
    async def get_cross_language_stub(self, address: str):
        """ç²å–è·¨èªè¨€æœå‹™å­˜æ ¹"""
        if not GRPC_AVAILABLE:
            return None
            
        stub_key = f"{address}_cross_lang"
        if stub_key not in self._stubs:
            channel = await self.get_channel(address)
            if channel:
                self._stubs[stub_key] = aiva_pb2_grpc.CrossLanguageServiceStub(channel)
        
        return self._stubs.get(stub_key)
    
    async def close_all(self):
        """é—œé–‰æ‰€æœ‰é€£æ¥"""
        for address, channel in self._channels.items():
            try:
                await channel.close()
                logger.info(f"ğŸ”Œ é—œé–‰ gRPC é€£æ¥: {address}")
            except Exception as e:
                logger.warning(f"âš ï¸  é—œé–‰é€£æ¥å¤±æ•—: {address} - {e}")
        
        self._channels.clear()
        self._stubs.clear()


class AIVAGRPCClient:
    """AIVA gRPC å®¢æˆ¶ç«¯"""
    
    def __init__(self, server_address: str = "localhost:50051"):
        self.server_address = server_address
        self.connection_manager = GRPCConnectionManager()
        self._health_checked = False
    
    async def _ensure_connection(self) -> bool:
        """ç¢ºä¿ gRPC é€£æ¥å¯ç”¨"""
        if not GRPC_AVAILABLE:
            logger.debug("gRPC ä¸å¯ç”¨ï¼Œå°‡ä½¿ç”¨å‚™ç”¨å”è­°")
            return False
            
        try:
            if not self._health_checked:
                # åŸ·è¡Œå¥åº·æª¢æŸ¥
                result = await self.health_check()
                self._health_checked = result.get("success", False)
            
            return self._health_checked
        except Exception as e:
            logger.warning(f"âš ï¸  gRPC é€£æ¥æª¢æŸ¥å¤±æ•—: {e}")
            return False
    
    async def create_task(self, task_config: Dict[str, Any]) -> Dict[str, Any]:
        """å‰µå»ºä»»å‹™"""
        try:
            if await self._ensure_connection():
                # ä½¿ç”¨ gRPC
                stub = await self.connection_manager.get_task_stub(self.server_address)
                if stub and aiva_pb2:
                    request = aiva_pb2.TaskConfig(
                        task_id=task_config.get("task_id", str(uuid.uuid4())),
                        task_type=task_config.get("task_type", "scan"),
                        target=aiva_pb2.Target(
                            url=task_config.get("target_url", ""),
                            host=task_config.get("host", ""),
                            port=task_config.get("port", 80)
                        ),
                        parameters=task_config.get("parameters", {}),
                        priority=task_config.get("priority", 5),
                        timeout=task_config.get("timeout", 300)
                    )
                    
                    response = await stub.CreateTask(request)
                    logger.info(f"âœ… gRPC ä»»å‹™å‰µå»ºæˆåŠŸ: {response.request_id}")
                    
                    return {
                        "success": response.success,
                        "task_id": response.request_id,
                        "protocol": "gRPC",
                        "result": dict(response.result) if hasattr(response, 'result') else {}
                    }
            
            # å‚™ç”¨: ä½¿ç”¨ MQ
            logger.info("ğŸ”„ gRPC ä¸å¯ç”¨ï¼Œä½¿ç”¨ MQ å‚™ç”¨å”è­°")
            message = message_broker.publish(
                topic=Topic.TASK_SCAN_START,
                payload=task_config,
                source_module=ModuleName.API_GATEWAY,
                target_module=ModuleName.SCAN,
                trace_id=str(uuid.uuid4())
            )
            
            return {
                "success": True,
                "task_id": task_config.get("task_id", str(uuid.uuid4())),
                "protocol": "MQ",
                "message_id": message.header.message_id
            }
            
        except Exception as e:
            logger.error(f"âŒ ä»»å‹™å‰µå»ºå¤±æ•—: {e}")
            return {"success": False, "error": str(e)}
    
    async def get_task_status(self, task_id: str) -> Dict[str, Any]:
        """ç²å–ä»»å‹™ç‹€æ…‹"""
        try:
            if await self._ensure_connection():
                # ä½¿ç”¨ gRPC
                stub = await self.connection_manager.get_task_stub(self.server_address)
                if stub and aiva_pb2:
                    request = aiva_pb2.AIVARequest(
                        request_id=task_id,
                        task="get_status",
                        parameters={},
                        trace_id=str(uuid.uuid4())
                    )
                    
                    response = await stub.GetTaskStatus(request)
                    logger.info(f"ğŸ“Š gRPC ä»»å‹™ç‹€æ…‹æŸ¥è©¢: {task_id}")
                    
                    return {
                        "success": True,
                        "task_id": response.task_id,
                        "status": response.status,
                        "protocol": "gRPC"
                    }
            
            # å‚™ç”¨: è¿”å›æ¨¡æ“¬ç‹€æ…‹
            logger.info("ğŸ”„ gRPC ä¸å¯ç”¨ï¼Œè¿”å›æ¨¡æ“¬ç‹€æ…‹")
            return {
                "success": True,
                "task_id": task_id,
                "status": "PENDING",
                "protocol": "MQ_FALLBACK"
            }
            
        except Exception as e:
            logger.error(f"âŒ ä»»å‹™ç‹€æ…‹æŸ¥è©¢å¤±æ•—: {e}")
            return {"success": False, "error": str(e)}
    
    async def cancel_task(self, task_id: str) -> Dict[str, Any]:
        """å–æ¶ˆä»»å‹™"""
        try:
            if await self._ensure_connection():
                # ä½¿ç”¨ gRPC
                stub = await self.connection_manager.get_task_stub(self.server_address)
                if stub and aiva_pb2:
                    request = aiva_pb2.AIVARequest(
                        request_id=task_id,
                        task="cancel",
                        parameters={},
                        trace_id=str(uuid.uuid4())
                    )
                    
                    response = await stub.CancelTask(request)
                    logger.info(f"âŒ gRPC ä»»å‹™å–æ¶ˆ: {task_id}")
                    
                    return {
                        "success": response.success,
                        "task_id": task_id,
                        "protocol": "gRPC"
                    }
            
            # å‚™ç”¨: ä½¿ç”¨ MQ
            logger.info("ğŸ”„ gRPC ä¸å¯ç”¨ï¼Œä½¿ç”¨ MQ å–æ¶ˆä»»å‹™")
            message_broker.publish(
                topic=Topic.COMMAND_TASK_CANCEL,
                payload={"task_id": task_id},
                source_module=ModuleName.API_GATEWAY,
                target_module=ModuleName.CORE,
                trace_id=str(uuid.uuid4())
            )
            
            return {
                "success": True,
                "task_id": task_id,
                "protocol": "MQ"
            }
            
        except Exception as e:
            logger.error(f"âŒ ä»»å‹™å–æ¶ˆå¤±æ•—: {e}")
            return {"success": False, "error": str(e)}
    
    async def stream_task_progress(self, task_id: str) -> AsyncIterator[Dict[str, Any]]:
        """ä¸²æµä»»å‹™é€²åº¦"""
        try:
            if await self._ensure_connection():
                # ä½¿ç”¨ gRPC
                stub = await self.connection_manager.get_task_stub(self.server_address)
                if stub and aiva_pb2:
                    request = aiva_pb2.AIVARequest(
                        request_id=task_id,
                        task="stream_progress",
                        parameters={},
                        trace_id=str(uuid.uuid4())
                    )
                    
                    logger.info(f"ğŸ“¡ é–‹å§‹ gRPC é€²åº¦ä¸²æµ: {task_id}")
                    
                    async for response in stub.StreamTaskProgress(request):
                        yield {
                            "task_id": task_id,
                            "status": dict(response.result).get("status", "UNKNOWN"),
                            "message": dict(response.result).get("message", ""),
                            "protocol": "gRPC",
                            "timestamp": response.timestamp
                        }
                    
                    logger.info(f"âœ… gRPC é€²åº¦ä¸²æµå®Œæˆ: {task_id}")
                    return
            
            # å‚™ç”¨: æ¨¡æ“¬é€²åº¦æ›´æ–°
            logger.info("ğŸ”„ gRPC ä¸å¯ç”¨ï¼Œä½¿ç”¨æ¨¡æ“¬é€²åº¦ä¸²æµ")
            progress_steps = [
                ("RUNNING", "ä»»å‹™é–‹å§‹åŸ·è¡Œ"),
                ("PROGRESS", "è™•ç†ä¸­... 50%"),
                ("COMPLETED", "ä»»å‹™å®Œæˆ")
            ]
            
            for status, message in progress_steps:
                yield {
                    "task_id": task_id,
                    "status": status,
                    "message": message,
                    "protocol": "SIMULATION",
                    "timestamp": datetime.now()
                }
                await asyncio.sleep(1)
                
        except Exception as e:
            logger.error(f"âŒ é€²åº¦ä¸²æµå¤±æ•—: {e}")
            yield {"success": False, "error": str(e)}
    
    async def execute_cross_language_task(self, task: str, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """åŸ·è¡Œè·¨èªè¨€ä»»å‹™"""
        try:
            if await self._ensure_connection():
                # ä½¿ç”¨ gRPC
                stub = await self.connection_manager.get_cross_language_stub(self.server_address)
                if stub and aiva_pb2:
                    request = aiva_pb2.AIVARequest(
                        request_id=str(uuid.uuid4()),
                        task=task,
                        parameters=parameters,
                        trace_id=str(uuid.uuid4())
                    )
                    
                    response = await stub.ExecuteTask(request)
                    logger.info(f"ğŸ”„ gRPC è·¨èªè¨€ä»»å‹™åŸ·è¡Œ: {task}")
                    
                    return {
                        "success": response.success,
                        "task": task,
                        "protocol": "gRPC",
                        "result": dict(response.result) if hasattr(response, 'result') else {}
                    }
            
            # å‚™ç”¨: ä½¿ç”¨ MQ
            logger.info("ğŸ”„ gRPC ä¸å¯ç”¨ï¼Œä½¿ç”¨ MQ åŸ·è¡Œè·¨èªè¨€ä»»å‹™")
            message = message_broker.publish(
                topic=Topic.TASK_SCAN_START,
                payload=parameters,
                source_module=ModuleName.API_GATEWAY,
                target_module=ModuleName.FUNCTION,
                trace_id=str(uuid.uuid4())
            )
            
            return {
                "success": True,
                "task": task,
                "protocol": "MQ",
                "message_id": message.header.message_id
            }
            
        except Exception as e:
            logger.error(f"âŒ è·¨èªè¨€ä»»å‹™åŸ·è¡Œå¤±æ•—: {e}")
            return {"success": False, "error": str(e)}
    
    async def health_check(self) -> Dict[str, Any]:
        """å¥åº·æª¢æŸ¥"""
        try:
            if not GRPC_AVAILABLE:
                return {"success": True, "protocol": "MQ_ONLY", "grpc_available": False}
                
            stub = await self.connection_manager.get_cross_language_stub(self.server_address)
            if stub and aiva_pb2:
                request = aiva_pb2.AIVARequest(
                    request_id=str(uuid.uuid4()),
                    task="health_check",
                    parameters={},
                    trace_id=str(uuid.uuid4())
                )
                
                response = await stub.HealthCheck(request)
                logger.info("ğŸ’“ gRPC å¥åº·æª¢æŸ¥é€šé")
                
                return {
                    "success": response.success,
                    "protocol": "gRPC",
                    "grpc_available": True,
                    "health_info": dict(response.result) if hasattr(response, 'result') else {}
                }
            
            return {"success": False, "protocol": "UNKNOWN", "grpc_available": False}
            
        except Exception as e:
            logger.warning(f"âš ï¸  gRPC å¥åº·æª¢æŸ¥å¤±æ•—: {e}")
            return {"success": False, "error": str(e), "grpc_available": False}
    
    async def get_service_info(self) -> Dict[str, Any]:
        """ç²å–æœå‹™è³‡è¨Š"""
        try:
            if await self._ensure_connection():
                stub = await self.connection_manager.get_cross_language_stub(self.server_address)
                if stub and aiva_pb2:
                    request = aiva_pb2.AIVARequest(
                        request_id=str(uuid.uuid4()),
                        task="get_service_info",
                        parameters={},
                        trace_id=str(uuid.uuid4())
                    )
                    
                    response = await stub.GetServiceInfo(request)
                    logger.info("â„¹ï¸ ç²å– gRPC æœå‹™è³‡è¨Š")
                    
                    return {
                        "success": response.success,
                        "protocol": "gRPC",
                        "service_info": dict(response.result) if hasattr(response, 'result') else {}
                    }
            
            # å‚™ç”¨è³‡è¨Š
            return {
                "success": True,
                "protocol": "MQ_FALLBACK",
                "service_info": {
                    "service_name": "AIVA Unified Client",
                    "protocols": ["MQ", "HTTP"],
                    "grpc_available": False
                }
            }
            
        except Exception as e:
            logger.error(f"âŒ ç²å–æœå‹™è³‡è¨Šå¤±æ•—: {e}")
            return {"success": False, "error": str(e)}
    
    async def close(self):
        """é—œé–‰å®¢æˆ¶ç«¯"""
        await self.connection_manager.close_all()
        logger.info("ğŸ”Œ gRPC å®¢æˆ¶ç«¯å·²é—œé–‰")


# å…¨åŸŸ gRPC å®¢æˆ¶ç«¯å¯¦ä¾‹  
grpc_client = AIVAGRPCClient()